### ISSUE TITLE
[selectors-5] add new selector immediate-adjecent-child

### ISSUE THREAD

**Issue by Meteor0id** [w3c/csswg-drafts#2568]
I am missing sort of the immediate sibling selector but for children instead of siblings.

Something like article ~> p

A theoretical article ~> p:first-of-type::first-letter would solve the following difficulty I encountered, for which I could come up with not other elegant method.

[https://jsfiddle.net/r0k0eoza/6/](https://jsfiddle.net/r0k0eoza/6/)
```

<div class="wrap_page">
<div class="some_other_wrapper">
<article>
  <head>
  <p>
  This text should have its first letter selected
  </p>
  <p>
  This text should NOT have its first letter selected
  </p>
  <main>
  <p>
  This text should NOT have its first letter selected
  </p>
  </main>
  <p>
  This text should NOT have its first letter selected
  </p>
  <p>
  This text should NOT have its first letter selected
  </p>
  <section>
    <p>
  This text should NOT have its first letter selected
  </p>
  </section>
  </head>
  <footer>
    <p>
  This text should NOT have its first letter selected
  </p>
  </footer>
</article>
</div>
<footer>
  <p>
  This text should NOT have its first letter selected
  </p>
</footer>
</div>
<article>
  <p>
   Just to be complete, this one should be selected
  </p>
</article>
```

`.wrap_main p:first-of-type::first-letter, article p:first-of-type::first-letter {
color: red;
}`

**Comment by jonjohnjohnson** [comment]
@Meteor0id  `article > :first-child:first-of-type::first-letter` gives your text the meaning it describes.

**Comment by Meteor0id** [comment]
@jonjohnjohnson 
wow, classic typo indeed to use head where header was meant, embarrasing.
article > :first-child::first-letter works in this very quick example I wrote here, but I am thinking about dynamicly generated pages, whith somewhat differentiating structure. My example is lacking, but I belive the issue still stands.

**Comment by jonjohnjohnson** [comment]
Can you update the example code showcasing the full example of what you mean by...

> dynamically generated pages, with somewhat differentiating structure

So we can better decipher if in fact current selector options cannot handle your request? Something to show/describe what you mean by "immediate adjacent child selector". Also reference [(AB)USING CSS3'S :NTH-CHILD SELECTOR TO INVENT NEW ONES](https://grack.com/blog/2015/01/09/abusing-css3-selectors/) and what is already in the works with [selectors-4](https://drafts.csswg.org/selectors-4/) for your use case.

**Comment by Loirooriol** [comment]
Just use `article > p:first-of-type::first-letter` instead of `article p:first-of-type::first-letter`? I don't get how your proposed `~>` would differ from `>`.

**Comment by Meteor0id** [comment]
I have been trying to summarise what I am missing the last 2 hours, but it's a lot to take in.
I will come back to this ticket as soon as I can set up a clear example to demonstrate exactly what I need.

For now I will try my best:
I am trying to set up a website using a framework. Each page the framework will handle has a different structure, but they all use the same main.css external style sheet. I am trying to add a rule in the stylesheet to select the very first paragraph element, but only the very first, and have its letter a larger font size. Think of a newspaper, in which each acticle opens with a stylised letter. The rules I use to style the letter are not relevant to this issue. The issue is selecting exclusively the first paragraph encountered on the page, without selecting any other paragraphs which follow as child or sibling or even in the parents of this first paragraph.
I do not know ahead of time what the parent element of the first paragraph will be, so folowing selector will not be an option
`.something > p:first-of-type { }`

Currently the selector would look something like
`body > *:first-of-type > p:first-of-type { }`
which is obviously not a valid statement.

`body p:first-of-type { }`
will not work either, as it will select not only the first overall <p> element on the page, but also many other <p> elements wich ar ethe first of their type in their own parent.

simply trying
`p:first-of-type { }`
also selects each first <p> as a child of any parent on the page, instead of just the absolute first.

Sibling selectors such as + and ~ are useless in this case, as I do not know to what element that first <p> will be a sibling ahead of time. Again, the stylesheet is used for a dynamicly generated page (inside a framework)

Maybe what I need is magic, but since I am looking for the first element meeting the right conditions which will be encountered while parsing from the top of the page, it initially seems to me it is technicly possible to devise a pseudo-element or combinator for this.

The idea I am trying to convey is very simulair to the difference between ~ and +, in which the former (~) selects any sibling of the named type and the latter (+) selects only a sibling which immediatly follows upon.
In my case I need something not for siblings, but for children. For example, I want the first <p> element which follows on <body>, looking at nested elements before looking at siblings.

**I am even confused whether I am looking for a combinator or a pseudo-element at this point**
Made enough of a fool of myself for today, so I will leave it at this for now.

If anyone managed to pick up on what I am looking for and knows whether it deseves further discussion please add a comment.

P.s.:
the issue stems from unexpected behavior of the pseudo-element :first-of-type (unexpected to me that is). I expected I could use it to select the very first element of a spesific type across the page, but what it really does is select each first element of given type of each element on the page.

I will not be sleeping.

**Comment by Meteor0id** [comment]
the issue is that :first-of-type is not exlusive to the first encountered within the scope, but rather to the first encountered of its own parent, allowing for multiple first-of-type to be matched within the scope.

**Comment by Loirooriol** [comment]
OK, I think the problem is that having `nth` pseudo-classes that count globally is bad for parallelization. So I suspect some browsers may be opposed to this, at least for the live profile.

**Comment by tabatkins** [comment]
Oh wait, nevermind, I was misreading. They are indeed asking for something larger.

@Meteor0id: Note that "the thing that's like + vs ~, but for children" already exists - it's `>` (versus " ", the descendant combinator).  What you're actually asking for didn't previously have an analog in either siblings *or* children; the new additions to `:nth-child()` do happen to bring this functionality into sibling lists, but we have no such way to search among the descendant list.  This is likely because it is, as @Loirooriol points out, dramatically more expensive to do so, making it somewhat unsuitable for the [live profile](https://drafts.csswg.org/selectors/#live-profile).  This is something you should instead just write a little bit of JS for: a treewalk using [`Element.matches()`](https://dom.spec.whatwg.org/#dom-element-matches) is easy and small to write.

**Comment by tabatkins** [comment]
For example:

```
function firstMatch(el, pred) {
  if(pred(el)) return el;
  for(const child of el.childElements) {
    const result = firstMatch(child);
    if(result) return result;
  }
}

const firstPOnPage = firstMatch(document.body, x=>x.matches("p"));
```

would find the first `p` element on the page.

**Comment by Loirooriol** [comment]
Or also `document.body.querySelector("p")`.

**Comment by tabatkins** [comment]
Ahahaha, yeah, that definitely works for this case. ^_^

**Comment by bkardell** [comment]
It seems to me that there isn't much common interest in this after 8 years, do we want to keep it open still? There were close solutions to this in 2018 even.

**Comment by tabatkins** [comment]
Yeah, I'll go ahead and reject this. The request (select the first element matching this selector, among the entire tree in tree order) is clear, but seems fairly niche in practice, and is solved pretty easily via a number of different methods that require just slightly more effort. (Always tag the container you're gonna look in so :first-of-type is useful, use a JS one-liner to tag the element in question, etc.) And making it react to updates is pricey, while the actual use-case seems to not need reactivity.
